File: scala/nucleotide-count/src/main/scala/Dna.scala
class DNA (chain: String) {

  type Count = Int
  type Nucleotide = Char
  type Nucleotides = Map[Nucleotide, Count]

  private val valid = List('A','T','C','G')
  private val defaults = valid.map(_ -> 0)
  private val empty = Map[Nucleotide, Count](defaults: _*)

  private val nucleotides = chain.foldLeft(empty) { (map, n) =>
    require(valid.exists(n == _))
    map + (n -> (map(n) + 1))
  }

  def count(n: Nucleotide): Count = {
    require(valid.exists(n == _))
    nucleotides(n)
  }
  def nucleotideCounts: Nucleotides = nucleotides
}

File: haskell/nucleotide-count/Dna.hs
module DNA(count, nucleotideCounts) where

import Data.Map

count :: Char -> String -> Int
count n chain = case Data.Map.lookup n (nucleotideCounts chain) of
                Nothing -> error "invalid nucleotide 'X'"
                Just x -> x

nucleotideCounts :: String -> Map Char Int
nucleotideCounts xs = nucleotideCounts' xs defaults

nucleotideCounts' :: String -> Map Char Int -> Map Char Int
nucleotideCounts' [] acc = acc
nucleotideCounts' (x:xs) acc = nucleotideCounts' xs (insert x next acc)
  where next = (succ (acc ! x))

defaults :: Map Char Int
defaults = fromList [('A',0),('C',0),('G',0),('T',0)]